File: base/tests/refinement_strategy/test_SubspaceRefinement.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright (C) 2008-today The SG++ project
# This file is part of the SG++ project. For conditions of distribution and
# use, please see the copyright notice provided with SG++ or at 
# sgpp.sparsegrids.org

import unittest
from pysgpp import Grid, HashRefinement, HashGridPoint, \
    SurplusRefinementFunctor, DataVector, SubspaceRefinement


class TestSubspaceRefinement(unittest.TestCase):



    def setUp(self):
        self.grid = Grid.createLinearGrid(2)  # a simple 2D grid
        self.grid.getGenerator().regular(3)  # max level 3 => 17 points
        self.HashGridStorage = self.grid.getStorage()


    def tearDown(self):
        del self.grid
        del self.HashGridStorage 
    
    
    def test_freeRefineSubspaceAnisotropic(self):
        """Refine Anisotropic subspace (x2)"""
        alpha = DataVector(self.grid.getSize())
        alpha.setAll(1.0)
        for i in [9, 10, 11, 12]:
            alpha[i] = 2.
        #refinement  stuff
        refinement = HashRefinement()
        decorator = SubspaceRefinement(refinement)
        # refine a single grid point each time
        functor = SurplusRefinementFunctor(alpha,1)
        decorator.free_refine(self.HashGridStorage,functor)
        for i in range(self.grid.getSize()):
            HashGridPoint = self.HashGridStorage.getPoint(i)

        self.assertEqual(self.grid.getSize(), 33)
        
        for i in range(self.grid.getSize()):
            HashGridPoint = self.HashGridStorage.getPoint(i)
            levelIndex = eval(HashGridPoint.toString())
            self.assertFalse(levelIndex[0] == 4)
            
            
    def test_freeRefineSubspaceIsotropic(self):
        """Refine the isotropic middle subspace"""
        alpha = DataVector(self.grid.getSize())
        alpha.setAll(1.0)
        for i in [13, 14, 15, 16]:
            alpha[i] = 2.
        #refinement  stuff
        refinement = HashRefinement()
        decorator = SubspaceRefinement(refinement)
        # refine a single grid point each time
        functor = SurplusRefinementFunctor(alpha,1)
        decorator.free_refine(self.HashGridStorage,functor)
        for i in range(self.grid.getSize()):
            HashGridPoint = self.HashGridStorage.getPoint(i)

        self.assertEqual(self.grid.getSize(), 33)
        
        for i in range(self.grid.getSize()):
            HashGridPoint = self.HashGridStorage.getPoint(i)
            levelIndex = eval(HashGridPoint.toString())
            self.assertFalse(levelIndex[0] == 4 or levelIndex[2] == 4)



if __name__ == '__main__':
    # import sys;sys.argv = ['', 'Test.testName']
    unittest.main()

File: base/tests/refinement_strategy/test_SubspaceRefinement.cpp
// Copyright (C) 2008-today The SG++ project
// This file is part of the SG++ project. For conditions of distribution and
// use, please see the copyright notice provided with SG++ or at
// sgpp.sparsegrids.org

#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>

#include <sgpp/base/datatypes/DataVector.hpp>
#include <sgpp/base/grid/generation/functors/SurplusRefinementFunctor.hpp>
#include <sgpp/base/grid/generation/hashmap/HashGenerator.hpp>
#include <sgpp/base/grid/generation/hashmap/HashRefinement.hpp>
#include <sgpp/base/grid/storage/hashmap/HashGridPoint.hpp>
#include <sgpp/base/grid/storage/hashmap/HashGridStorage.hpp>
#include <sgpp/base/grid/generation/refinement_strategy/SubspaceRefinement.hpp>

using sgpp::base::DataVector;
using sgpp::base::HashGenerator;
using sgpp::base::HashGridPoint;
using sgpp::base::HashGridStorage;
using sgpp::base::HashRefinement;
using sgpp::base::SubspaceRefinement;
using sgpp::base::SurplusRefinementFunctor;

BOOST_AUTO_TEST_SUITE(TestSubspaceRefinement)

BOOST_AUTO_TEST_CASE(testFreeRefineTrivial) {
  HashGridStorage storage(2);
  HashGenerator generator;

  generator.regular(storage, 1);

  DataVector datavector(1);
  datavector[0] = 1.0;

  SurplusRefinementFunctor functor(datavector);
  HashRefinement* hash_refinement = new HashRefinement();
  SubspaceRefinement subspace_refinement(hash_refinement);

  subspace_refinement.free_refine(storage, functor);

  BOOST_CHECK_EQUAL(storage.getSize(), 5);

  delete hash_refinement;
}

BOOST_AUTO_TEST_CASE(testFreeRefineSubspaceAnisotropic) {
  HashGridStorage storage(2);
  HashGenerator generator;

  generator.regular(storage, 3);

  DataVector data_vector(storage.getSize());
  data_vector[9] = 2.0;

  SurplusRefinementFunctor functor(data_vector, 1);
  HashRefinement* hash_refinement = new HashRefinement();

  SubspaceRefinement subspace_refinement(hash_refinement);

  subspace_refinement.free_refine(storage, functor);

  BOOST_CHECK_EQUAL(storage.getSize(), 33);

  for (size_t i = 0; i < storage.getSize(); i++) {
    HashGridPoint& index = storage.getPoint(i);
    BOOST_CHECK((index.getIndex(0) == 4) == false);
  }

  delete hash_refinement;
}

BOOST_AUTO_TEST_CASE(testFreeRefineSubspaceIsotropic) {
  HashGridStorage storage(2);
  HashGenerator generator;

  generator.regular(storage, 3);

  DataVector data_vector(storage.getSize());
  data_vector[13] = 2.0;

  SurplusRefinementFunctor functor(data_vector, 1);
  HashRefinement* hash_refinement = new HashRefinement();

  SubspaceRefinement subspace_refinement(hash_refinement);

  subspace_refinement.free_refine(storage, functor);

  BOOST_CHECK_EQUAL(storage.getSize(), 33);

  for (size_t i = 0; i < storage.getSize(); i++) {
    HashGridPoint& index = storage.getPoint(i);
    BOOST_CHECK((index.getIndex(0) == 4 || index.getIndex(1) == 4) == false);
  }

  delete hash_refinement;
}

BOOST_AUTO_TEST_SUITE_END()
